src/inc/til/u8u16convert.h

```diff
@@ -306,25 +89,117 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
         }
     }
 
+#pragma warning(push)
+#pragma warning(disable : 26429 26446 26459 26481 26482 26485) // use not_null, subscript operator, use span, pointer arithmetic, dynamic array indexing, array to pointer decay
     // Routine Description:
     // - Takes a UTF-8 string, complements and/or caches partials, and performs the conversion to UTF-16.
     // Arguments:
     // - in - UTF-8 string to be converted
     // - out - reference to the resulting UTF-16 string
-    // - state - reference to a til::u8state class holding the status of the current partials handling
+    // - state - reference to a til::u8state holding the status of the current partials handling
     // Return Value:
     // - S_OK          - the conversion succeeded
     // - E_OUTOFMEMORY - the function failed to allocate memory for the resulting string
     // - E_ABORT       - the resulting string length would exceed the upper boundary of an int and thus, the conversion was aborted before the conversion has been completed
     // - E_UNEXPECTED  - an unexpected error occurred
-    template<class inT, class outT>
-    [[nodiscard]] typename std::enable_if<std::is_same<typename inT::value_type, char>::value && std::is_same<typename outT::value_type, wchar_t>::value, HRESULT>::type
-    u8u16(const inT in, outT& out, u8state& state) noexcept
+    template<class outT>
+    [[nodiscard]] HRESULT u8u16(const std::string_view& in, outT& out, u8state& state) noexcept
     {
-        std::string_view sv{};
-        RETURN_IF_FAILED(state(std::string_view{ in }, sv));
-        return til::u8u16(sv, out);
+        try
+        {
+            out.clear();
+            RETURN_HR_IF(S_OK, in.empty());
+
+            int capa16{};
+            // The worst ratio of UTF-8 code units to UTF-16 code units is 1 to 1 if UTF-8 consists of ASCII only.
+            RETURN_HR_IF(E_ABORT, !base::CheckAdd(in.length(), state.have).AssignIfValid(&capa16));
+
+            out.resize(gsl::narrow_cast<size_t>(capa16));
+            auto len8{ gsl::narrow_cast<int>(in.length()) };
+            int len16{};
+            auto cursor8{ in.data() };
+            if (state.have)
+            {
+                if (state.want > len8) // we still didn't get enough data to complete the code point, however this is not an error
+                {
+                    std::move(cursor8, cursor8 + len8, state.partials + state.have);

```

Comment:
Eventually I continued with std::move. Using memcpy we trigger a signed/unsigned conflict. This would have required another type cast. We have quite some of them already 😆

Reply:
@lhecker Oh, the missing `out.clear()` is a candidate for the failing UT. I already resized the string. IIRC last time I tried to only `reserve()` more capacity it crashed. Maybe `data()` still points to a short static buffer in this case.